[aotd] Support HOP effects in backward

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132755
* __->__ #132638

Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

IvanKobzarev committed:
[aotd] Support HOP effects in backward

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -292,6 +292,10 @@ def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=Fals
         self._tokens: Dict[Any, torch.Tensor] = {}
 
+        # Filled after forward tracing is done, see on_forward_tracing_end
+        self._tokens_forward_output: Dict[Any, torch.Tensor] = {}
```
just a nit, but right now this PR "violates invariants" in both directions:
(1) the `create_joint()` logic needs to manually mess around with tokens from `FunctionalTensorMode`
(2) `FunctionalTensorMode` needs to be aware of a fw/bw context and treat tokens separately between the two

If we are already going to have to live with (1) (the joint-tracing code knowing about effect tokens), I would rather not also do (2). The easiest way to do that is probably just to have the `create_joint` function manually grab `mode._tokens` at the beginning vs end of the backward, and diff the two tokens lists to know which tokens were only part of the backward.

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -230,6 +230,12 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
+                torch._C._TorchDispatchModeKey.FUNCTIONAL
+            )
+            if functional_tensor_mode is not None:
+                functional_tensor_mode.on_forward_tracing_end()
```
I think I was expecting that you would actually have to tweak the outputs of `joint_fn()` a bit. Right now, we pass in tokens as extra tensor inputs to `joint_fn`, and it returns more tokens as extra outputs of the joint. But if we want to signal to the partitioner that some output tokens should be "forward graph outputs" vs "backward graph outputs", I think we might need to do something like this?
```
        return outs + fw_token_outs, [
            next(backward_out_iter) if i else None for i in inputs_needs_grads
        ] + bw_token_outs
```

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -376,7 +376,7 @@ def add_hooks(module, config):
     def test_effectful_custom_op_with_subclass(self):
         with torch.library._scoped_library("_mylib", "FRAGMENT") as lib:
-            lib.define("foo(Tensor x) -> Tensor")
+            lib.define("zoo(Tensor x) -> Tensor")
```
Can you add a test where we have an regular operator whose backward calls an operator that has a side effect?

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -532,6 +532,110 @@ def forward(self, tangents_1, tangents_2):
+                return a.clone()
+
+            def foo_bwd(ctx, grad):
+                return torch.ops._mylib.foo(grad)
```
can you also have this op increment a counter? That way we can test that the counter was actually incremented once in the forward and again in the backward

bdhirsh commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -532,6 +532,110 @@ def forward(self, tangents_1, tangents_2):
+                _deregister_effectful_op(torch.ops._mylib.foo.default)
+
+    @skipIfNoDynamoSupport
+    def test_regular_effectful_op_only_in_backward(self):
```
An expecttest of the fw and bw graphs would be nice for this test too (so we can see in the graph that the fw graph takes in an effect token and saves if for bw, and the bw actually uses said effect token in the effectful op)

bdhirsh commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -532,6 +532,110 @@ def forward(self, tangents_1, tangents_2):
+            _deregister_effectful_op(torch.ops.aten.cos.default)
+
+    @skipIfNoDynamoSupport
+    def test_regular_effectful_op_in_forward_and_backward(self):
```
hmm what is this testing differently from the above tests?

IvanKobzarev replied:
This one is testing regular effectful op (aten.cos) in forward and backward 
The previous tests:
 - custom op in forward and backward
 - regular in backward only

bdhirsh commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -234,59 +237,116 @@ def maybe_to_fresh_input(idx, t, meta):
 
-    fw_module.recompile()
+    # No need to update fw_metadata.num_forward_returns and fw_metadata.num_forward as num_tokens are not part of it.
+    # As CompiledFunction.forward runs function wrapped in EffectTokensWrapper, that will toss out output tokens.
```
This comment is confusing to me because in particular, in the path where we call `unlift_tokens` (when `config.unlift_effect_tokens` is set), the EffectTokensWrapper doesn't actually have to do anything at runtime since there are no extra token input/output runtime arguments

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -1734,6 +1736,9 @@ def classify_nodes(joint_module):
             if node.op == "placeholder" and "tangents" in node.target:
                 required_bw_nodes.add(node)
+            elif _has_tag_must_be_in_backward(node):
+                required_bw_nodes.add(node)
```
Is this fixing a bug in the partitioner that is technically independent of effect tokens? (it would be nice if we could add a test in a fresh PR that involves mutations in the backward in some way that is fixed by this, independently of the changes for getting effect-tokens-in-backward to work). If this is a pain though then np.

IvanKobzarev replied:
Yes, will be good to have a separate PR with test cases, but can be difficult to make the needed graph with mutations in backward.

IvanKobzarev replied:
Realized that the bug happens, when is_must_be_in_backward node is picked as a saved_value by partitioner. We do not have control to specify to partitioner which nodes to pick, so the test will be flaky. Once we have effectful ops as "must_be_in_backward" - it will be easier to make a test (like added here with_effects_only_in_backward)

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -428,10 +433,17 @@ def unwrap(x):
-            return handle_effects(
-                self._allow_token_discovery, self._tokens, func, args, kwargs
+            from torch._functorch._aot_autograd.traced_function_transforms import (
+                set_partitioner_tag,
```
Right now you are directly using this context manager here in the `FunctionalTensor.__torch_dispatch__` code, because you are trying to carefully set the meta field for **only** any `with_effects()` ops that are called in the backward.

It feels bit bad that `FunctionalTensor` has to know about this code. Instead, what if we:

(1) agree that **all** nodes that are invoked as part of the `autograd.grad()` part of the joint function will get this metadata (basically just a bool that specifies "is_backward")

(2) the partitioner will just have access to this metadata for all backward nodes (even if it only really needs it for mutable ops / effectful ops)

bdhirsh commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -291,6 +298,27 @@ def aot_dispatch_autograd_graph(
+    # If backward graph does use effectful ops with Tokens,
+    # we have to add additional token argument to the joint graph manually after tracing.
+    if fw_metadata.num_backward_discovered_tokens > 0:
+        add_discovered_token_in_backward_as_input(
```
Hmm. Extra graph pass here definitely works, although in general I am a bit hesitant about adding extra graph passes in AOTDispatcher (my mental model today is that our steps are (1) metadata collection, (2) trace a joint graph, (3) partition, (4) compile. And that the graph we got right after joint tracing might not be "correct" and is fixed up elsewhere feels nonlocal/hard to reason about).

I was imagining that instead of a fresh graph pass, we would add the new tokens as extra arguments **at the same time that we are in the middle of tracing**. 

One way of doing this is (which on second thought, might take some... experimentation) would be to update the `handle_effects` ([code](https://github.com/pytorch/pytorch/blob/32e7d284bdbafa6a2ca50b8b9519b5b3e09bdeae/torch/_higher_order_ops/effects.py#L197)) (called by`FunctionalTensor.__torch_dispatch__()`. If we are currently proxy-tracing while this code is running, then do some graph surgery on the fly to add the newly created token as an extra primal. Something like:
```
maybe_mode = torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY)
if maybe_mode:
    # we are in the middle of the tracing phase of AOTDispatcher
    graph  = maybe_mode.tracer.graph
    ...
    # grab the `primals` placeholder node off of `graph` and add the token tensor as an extra graph arg,
    # similar to what you did in `add_discovered_token_in_backward_as_input`
    # but only need to do for a single token
    # probably also need to mess with the pytree stuff...
    from torch.fx.experimental.proxy_tensor import track_tensor_tree, disable_proxy_modes_tracing
    with disable_proxy_modes_tracing():
         token = torch.tensor([])
     # hook up the newly created token to the currently-traced-graph's list of proxies
     token_proxy = maybe_m.tracer.create_proxy("placeholder", "primals_N", (), {}, name="primals_N")
     track_tensor_tree(token, token_proxy, constant=None, tracer=maybe_m.tracer)
```

Happy to chat more about this next week though

IvanKobzarev replied:
Removed the pass, but we still need num_backward_discovered_tokens to add additional arguments at runtime for aot_eager.

bdhirsh commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -291,6 +298,27 @@ def aot_dispatch_autograd_graph(
+            aot_graphs_effects_log.debug(
+                "%s",
+                lazy_format_graph_code(
+                    "aot_config id: %s, fw_metadata=%s, Joint graph after adding effect tokens discovered in backward",
```
tbh, I don't think that the log is necessary, since the existing `aot_graphs` logging will log the final joint and fw/bw graphs after effect tokens have been added anyway (also, this won't matter if we add the extra effect tokens are *during* tracing, from the other comment)

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +680,241 @@ def forward(self, arg0_1, arg1_1):
+            for backend in ["CPU", "CUDA", "Meta"]:
+                lib.impl("foo", foo_impl, backend)
+
+            if torch._C._dispatch_has_kernel_for_dispatch_key(
+                "_mylib::foo", "Autograd"
+            ):
+                self.skipTest(
+                    "Double registration of Autograd kernel for test custom op"
+                )
```
We shouldn't need this code

IvanKobzarev replied:
Yes, it was failing only on CI in parallel execution, wanted to debug this later. I think lib=lib should fix this, will add it.

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +680,241 @@ def forward(self, arg0_1, arg1_1):
+                    "Double registration of Autograd kernel for test custom op"
+                )
+
+            torch.library.register_autograd("_mylib::foo", foo_bwd)
```
Need to pass lib=lib to actually get the effects of _scoped_library

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +680,241 @@ def forward(self, arg0_1, arg1_1):
+                _register_effectful_op,
+            )
+
+            _register_effectful_op(torch.ops._mylib.foo.default, _EffectType.ORDERED)
```
the SIDE_EFFECTS dict should probably be a WeakKeyDictionary -- the operator can go out of scope

IvanKobzarev replied:
Will add TODO to SIDE_EFFECTS

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +680,241 @@ def forward(self, arg0_1, arg1_1):
+                        self.assertExpectedInline(
+                            fw_graph.code.strip(),
+                            """\
+def forward(self, primals_1, primals_2, primals_3):
+    with_effects = torch.ops.higher_order.with_effects(primals_1, torch.ops._mylib.foo.default, primals_2);  primals_1 = primals_2 = None
+    getitem = with_effects[0]
+    getitem_1 = with_effects[1];  with_effects = None
+    add = torch.ops.aten.add.Tensor(getitem_1, primals_3);  getitem_1 = primals_3 = None
+    return (getitem, add, getitem)""",
```
It looks like we're threading the token from the forward to the backward. Is there a reason why we didn't make a separate token for the backward? This would avoid some of the problems with `must_be_in_backward`, because the token would need to be in the backward.

bdhirsh replied:
Ivan and I talked a bit offline about it. I thought that the `must_be_in_backward` changes were kind of necessary anyway, if/when we have any other mutations that can show up in the middle of the backward graph (e.g. from FSDP), that we need to ensure the partitioner does not move to the forward.

We could move the backward tokens into tangents too - it just saves us from having to add the extra wrapper logic at runtime in the backward to pass in extra token inputs, and deal with the corresponding extra indexing logic (although we still need the wrapper at the end of the backward to remove the tokens in either case). Let me know if given the above you still have an preference on "backward tokens as saved activations vs tangents"

bdhirsh replied:
> This approach is also consistent with our future "pre-dispatch trace the backward and specialize later" plan -- it's possible that we won't see the effect token until we specialize, and at that point it is too late to change the forward graph.

oh that is true, that in "pre-dispatch backward graph" world we will probably want the effect token runtime logic to live entirely in the backward graph / as tangents

zou3519 replied:
> Maybe a degenerate case 😛 

Can DTensor issue wait_tensor calls? If so, then not a degenerate case. This is similar to Tensor subclassing issuing intermediate mutations in their torch_dispatch for functional operators, we put functionalization underneath subclasses because we were worried about that, right?

>  Let me know if given the above you still have an preference on "backward tokens as saved activations vs tangents"

I still prefer the tokens to be in the backward. Ignoring the need to beef up  "must_be_in_backward", putting the tokens in the backward feels like the morally correct solution (the backward graph is technically a separate graph from the forward graph) and is more composable with what we want to do in the future. How difficult would it be to do this / how much additional time would we spend on it?

bdhirsh replied:
I'll let @IvanKobzarev comment, but you're definitely right that the case I mentioned above is probably not very degenerate (e.g. if you manage to have a fw/bw graph where the only DTensor is coming from tangents, and so unless the tangents are DTensors then there will otherwise be no collectives in the backward).

On the partitioner changes, I could imagine us needing them if/when `set_()` and friends show up in the backward graph? Although I'm not sure if we should add that "fix" pre-emptively or wait for this case to show up in FSDP2 tracing to get a good test case.

zou3519 reviewed (commented):
This PR is kind of big, is it possible to split it?

My main question is around how we're threading the tokens from the forward to the backward. Originally I expected that the backward would accept its own tokens (that don't come from the forward) and that that would save us from needing to add logic to force side-effectful operators into the backward graph. This approach is also consistent with our future "pre-dispatch trace the backward and specialize later" plan -- it's possible that we won't see the effect token until we specialize, and at that point it is too late to change the forward graph.

bdhirsh commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -234,57 +237,173 @@ def maybe_to_fresh_input(idx, t, meta):
+    return node.op == "call_function" and node.target.__name__ == "with_effects"
+
+
+def has_with_effects_ancestors(node):
```
unused?

IvanKobzarev replied:
yes :)

IvanKobzarev commented:
> This PR is kind of big, is it possible to split it?
> 
> My main question is around how we're threading the tokens from the forward to the backward. Originally I expected that the backward would accept its own tokens (that don't come from the forward) and that that would save us from needing to add logic to force side-effectful operators into the backward graph. This approach is also consistent with our future "pre-dispatch trace the backward and specialize later" plan -- it's possible that we won't see the effect token until we specialize, and at that point it is too late to change the forward graph.

1/ Yes, it's possible to split partitioner changes to a separate PR, but hard to write tests for them without effectful_ops with must_be_in_backward, especially if tokens are not saved from forward. Will split it.

2/ Will make backward tokens independent from forward.

IvanKobzarev commented:
Splitted partitioner must_be_in_backward fixes to https://github.com/pytorch/pytorch/pull/134002

IvanKobzarev commented:
Splitted change of tokens dtype to torch.bool - https://github.com/pytorch/pytorch/pull/134003

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -235,22 +236,45 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
-                        needed_outs,
-                        grad_primals,
-                        allow_unused=True,
+            with contextlib.ExitStack() as context_stack:
```
Why the switch to `contextlib.ExitStack()` over an explicit context manager?
```
with fx_traceback.preserve_node_meta(), set_partitioner_tag(...):
    ....
```

IvanKobzarev replied:
Yeah, in the original version with saved values there were more control-logic, now can be simplified.

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -299,6 +299,9 @@ def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=Fals
         self._tokens: Dict[Any, torch.Tensor] = {}
 
+        # Filled after forward tracing.
+        self._tokens_forward_output: Dict[Any, torch.Tensor] = {}
```
Hmm I was going to say it would be nice if we don't need to add any state at all to `FunctionalTensorMode`. But I guess if you remove this, you'll need another way to plumb the list of forward-only tokens up the `joint_fn()` to the `handle_effect_tokens_fn`, so maybe sticking a field here isn't so bad.

bdhirsh replied:
oh, looking at your code changes in `handle_effect_tokens_fn`, you technically don't need to do this: you can just grab the entire list of tokens after tracing the joint, and assume that if `len(functional_mode._tokens) > len(fw_metadata.tokens)`, the extra tokens on our functional mode correspond to backward tokens.

If you want to keep the assert it that sounds ok, but if so I would add a comment here describing the situation (that the only reason we're storing this extra state on `FunctionalTensorMode` is to more easily add asserts about the number of fw vs bw tokens).

IvanKobzarev replied:
If we keep them inside _tokens for backward, they will be reused in backward.
So we need some place to store them and to add to the outs of runtime.
Do you mean to add just [None] * num_tokens instead of real result values?

bdhirsh replied:
resolved offline - we're using `FunctionalTensorMode` here as a global object to stash the forward tokens, so that later in the token-wrapper-fn we know which tokens are fw_outs and bw_outs (a comment describing the situation in `FunctionTensorMode` would be great)

bdhirsh commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -234,57 +237,151 @@ def maybe_to_fresh_input(idx, t, meta):
 
-def unlift_tokens(fw_module, fw_metadata):
+def is_with_effects(node):
+    return node.op == "call_function" and node.target.__name__ == "with_effects"
```
nit: instead of string comparison, you can do `node.target is torch.ops.higher_order.with_effects`

bdhirsh commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -484,13 +490,18 @@ def aot_dispatch_autograd(
+                + inner_meta.num_outputs_rng_offset
+                + num_backward_tokens
             )
             bw_outs_no_rng = bw_outs
```
nit: rename to `bw_outs_no_rng_or_tokens`?

bdhirsh commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +673,234 @@ def forward(self, arg0_1, arg1_1):
+                    ref_out = fn(*ref_ins)
+                    ref_out.sum().backward()
+
+                    compiled_fn = torch.compile(fn, backend="aot_eager", fullgraph=True)
```
can these tests run under both settings of the `unlift_tokens` config? (so we're properly testing the unlifting / no unlifting path separately, both for the case where tokens appear in both fw and bw and the case where tokens only appear in the bw)

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1771,9 +1777,15 @@ def backward(ctx, *flat_args):
+                    _l -= roffset
+                    _r = -roffset
+
+                grad_output_types = [type(x) for x in all_args[_l:_r]]
```
ok what if.... instead of all of this, we just set `grad_output_types = [type(x) for x in flat_bw_args_with_grads]` (right before we `del flat_bw_args_with_grads` and lose access to it)?

All of this indexing hurts my brain :p

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1855,7 +1867,9 @@ def get_types_for_tangents(tangents):
                     )
-                    tangents_start_idx = len(all_args) - len_tangents - len(rng_args)
+                    tangents_start_idx = (
+                        len(all_args) - len_tangents - len(rng_args) - len(bw_tokens)
```
just as a note to myself, the full list of backward graph input args is:
`(ctx.symints, ctx.saved_tensors, tangents, rng_args, bw_tokens)`

To be extra defensive, an you assert that the computation here is gives you the same result as `len(ctx.symints) + len(ctx.saved_tensors)`?

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1968,6 +1982,16 @@ def call_compiled_backward():
+                    if num_bw_tokens > 0:
+                        out_tokens = out[-num_bw_tokens:]
+                        for t in out_tokens:
+                            assert t.numel() == 0
```
Just a thought: one (annoying) thing we need to balance in AOTDispatcher at runtime is having enough useful asserts without hurting the hot-path runtime performance too much (which you might see as part of https://github.com/pytorch/pytorch/issues/129457)

`out = out[:-num_bw_tokens]` will materialize a fresh list, so this will be more expensive on backward graphs that have a large number of outputs (empirically, we have internal models with > 100 graph inputs/outputs).

We are doing similarly egregious things in other parts of the runtime wrapper, so I'm ok with leaving this list slicing here for now, but definitely something to think about.

IvanKobzarev replied:
Thanks for catching, agree, will remove.

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -235,22 +236,45 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
-                        allow_unused=True,
+                )
+                if functional_tensor_mode is not None:
+                    # Prevent partitioner from moving effectful ops happened in backward to forward for side-effects correctness.
```
nit: this comment applies to your use of `set_partitioner_tag` further down, not to the setting of `_tokens_forward_outputs`

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -649,13 +673,13 @@ def inner_fn(*args):
                 args = (args[0][num_tokens:], *args[1:])
             else:
                 tokens = args[:num_tokens]
+                assert all(token.numel() == 0 for token in tokens)
```
? (why did we move + double the assert)

IvanKobzarev replied:
I would have to declare `tokens=[]` outside, type checker was not happy,  giving lints about specifying type of that empty list :)

bdhirsh reviewed (commented):
overall looks good, left several comments

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1730,20 +1730,21 @@ def backward(ctx, *flat_args):
                 # - note: donated buffer logic requires (*ctx.symints, *ctx.saved_tensors) showing up first
                 #   in the bw output order.
+
+                # Every dereference of ctx.saved_tensors incurs saved_tensors_hooks calls
```
ah good catch. This is also probably... the right thing to do anyway (we don't want to invoke the saved tensors hooks multiple times when someone uses them over compile)

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -348,6 +365,14 @@ def set_partitioner_tag(tag: str):
+
+
+def set_partitioner_tag_must_be_in_backward():
+    return set_partitioner_tag("must_be_in_backward")
```
hmm I figured we would just unify thesse two tags into one (since in both cases they're just telling the partitioner "please keep this op in the backward"). any reason to keep them separate?

IvanKobzarev replied:
"in_backward" is to mark all backward nodes, which is by default allowed to be moved by partitioner for compute in forward.
"must_be_in_backward" is to prevent this move from backward to forward.

In future we may have more cases, when  we want to keep some particular ops +  "is_backward" condition in backward, but marking them as "must_be_in_backward" will be not handy.

bdhirsh reviewed (approved):
minor last comments but lgtm!

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Support HOP effects in backward"


Support of effectful operations in backward:

1/ AOTD collects metadata from forward fn only, so we can have usage of effectful ops in backward, that were not used in forward => Allowing tokens discovery during joint function .

FunctionalTensorMode holds _tokens, in Joint function after tracing forward we memoize _tokens as `_tokens_forward_output`.

2/ Tokens are added as primals inputs (forward) in EffectTokensWrapper. 
Tokens that will be used in backward are in partitioner saved values. We do not have control on which positions they are saved in forward outputs.

2/ If new tokens discovered in backward after tracing joint_fn, the result graph will be manually added in the end of primals.
_aot_autograd/utils.py

3/ All effectful ops during backward are marked with 'must_be_in_backward' partitioner_tag, to prevent partiitoner to place them in forward.

For that functional_tensor_mode got new optional state `self._effects_partitioner_tag` for effectful ops, to set after tracing forward.

There are additional changes in partitioner to improve functionality of 'must_be_in_backward'

4/ Unlift tokens now should run for both forward and backward.
- As saved for backward tokens are placed on non static places - we identify input and output tokens to erase, by input and output of `with_effects` operation
- In forward we can have input tokens, discovered in backward, that are not used in with_effects ops in forward, but saved for backward. We identify them by position in forward inputs.

5/ Adding aot debug logging for graphs before unlifting and before adding additional primal for backward tokens.

Tests:
```
python test/higher_order_ops/test_with_effects.py
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev commented:
@pytorchbot merge

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -662,6 +674,234 @@ def forward(self, arg0_1, arg1_1):
+                    else:
+                        raise NotImplementedError
+            finally:
+                _deregister_effectful_op(torch.ops._mylib.foo.default)
```
mm the side effect registration should be tied to a library object

IvanKobzarev commented on torch/_higher_order_ops/effects.py:
```diff
@@ -23,6 +23,7 @@ class _EffectType(Enum):
 OpType = Union[torch._ops.HigherOrderOperator, torch._ops.OpOverload]
 
 
+# TODO(ivankobzarev): Make SIDE_EFFECTS dictionary WeakKeyDictionary as operator can go out of scope
```
Yes, will prepare a PR soon

IvanKobzarev replied:
https://github.com/pytorch/pytorch/pull/134470
